### PR TITLE
[sw] Add :memory_perftest to //sw/device/lib/base

### PR DESCRIPTION
**Pull request by dmcardle**
This on-device functest exercises memory functions like memcpy, memset,
etc. It uses RISC-V's mcycle counter to measure performance. Each
sub-test can run a configurable number of times, after which the
perftest computes summary statistics and prints to stdout.

Issue #13471

Signed-off-by: Dan McArdle <dmcardle@google.com>

### CODE REVIEW

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:166]: We generally try to avoid using `__attribute__` (and other double-underscore things) directly.

Create a macro for it in `sw/device/lib/base/macros.h` (there's already OT_ALWAYS_INLINE.  Seems like OT_NEVER_INLINE is needed for symmetry).

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:149]: Does `inline` do anything here?  You're taking the address of this function when you set up the table of benchmarks later, so this function cannot be inline for those cases.

It seems like `inline` has been added to everything.  I _think_ the C compiler is going to make its own decisions about whether to inline functions based on the optimization level; I'm not certain the `inline` keyword is useful for these functions.

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:257]: Given the setup for the `memcmp` family of functions, isn't memcmp likely to exit after comparing only a byte or two?  To test memcmp on long strides you're going to need buffers that are filled with the same sequence.

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:156]: Have a look at `sw/device/testing/rand_testutils.c`.

I'd say we should enhance those routines to allow seeding and  use them to generate the numbers used for this sequence.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:211]: Nice! I didn't realize it was returning the number of bytes printed, that's perfect.

I tried using the width specifier before writing this garbage, but it didn't work. Filed a bug to deal with this :) https://github.com/lowRISC/opentitan/issues/13873

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:166]: Great, that makes sense.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:149]: > Does inline do anything here? You're taking the address of this function when you set up the table of benchmarks later, so this function cannot be inline for those cases.

Nope, but it made sense before I refactored this and started taking the address of the function. Good catch.

> It seems like inline has been added to everything. I think the C compiler is going to make its own decisions about whether to inline functions based on the optimization level; I'm not certain the inline keyword is useful for these functions.

Maybe I was a bit overzealous :P. I was playing with `objdump` and noticed that `inline` was having an effect, so I tried to specify it when applicable.

Playing around in Compiler Explorer [^0] confirms your suspicion about optimization and inlining. I saw that when `-O2` is specified, both `inline` and not-necessarily-`inline` static functions are inlined. Without any `-O` flag, they are not inlined.

On that note, I guess we should be specifying an optimization level in the new functest's cflags. Not much point in perftesting code that is not optimized the same way it would be in production.

[^0]:  https://godbolt.org/z/dbxGvhMc5

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:257]: Thanks for pointing that out, I haven't really paid attention to the quality of the inputs after memset. For memcmp-like functions, I think it makes sense to copy the buffers, but make a single mutation in various locations.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:156]: Aha! I was looking for an RNG!

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:257]: I'm going to run memcmp and memrcmp on all zeroes for now, but I'm leaving a TODO that I'd like to resolve before we merge this.

**Comment by dmcardle** [comment]
Partially resolved @cfrantz's comments. I still need to figure out better inputs for mem[r]cmp and mem[r]chr.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [inline sw/device/lib/base/memory_perftest.c:44]: Oops, I need to update this comment. Stats are computed on `uint64_t` now, but I still don't have a nice way to print ints larger than 32 bits.

File: sw/device/lib/base/memory_perftest.c
Comment (alphan) [inline sw/device/lib/base/memory_perftest.c:46]: IMHO we can leave this out since at this level we don't expect a lot of variance, do we?

File: sw/device/lib/base/memory_perftest.c
Comment (alphan) [inline sw/device/lib/base/memory_perftest.c:98]: Please see my comment on `num_reps`. If you agree we can park this in a gist until we need it.

**Comment by alphan** [comment]
Thanks @dmcardle, some minor comments. I'm not sure if we need the statistics part though since this will practically run on bare metal. Did you have a particular reason in mind?

File: sw/device/lib/base/macros.h
Comment (dmcardle) [reply sw/device/lib/base/macros.h:48]: Good to know! Added a TODO referencing that PR.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:55]: I would have completely omitted the messages, but apparently that's not a thing in C11! I'll add better strings than `""`.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:96]: Neat! I didn't realize that existed. I guess I didn't need the ifdef — is it safe to assume we're running on Ibex?

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:49]: Yes! Done.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:318]: I haven't uploaded a PR yet, but I'm working on a correctness-focused set of unit tests. I figured this would be more focused on measuring the performance.

I suppose I could encode expectations in `perf_test_t`. Do you think that's worth doing?

File: sw/device/lib/base/memory_perftest.c
Comment (alphan) [reply sw/device/lib/base/memory_perftest.c:96]: Yes, this is a functest that should run only on the target platform. When we want to run the same test on different platforms, we can introduce different implementations.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:36]: I omitted the context parameter because it would be unused, but I don't feel strongly. Would you prefer I add it as a matter of style?

File: sw/device/lib/base/memory_perftest.c
Comment (alphan) [reply sw/device/lib/base/memory_perftest.c:318]: I think so. We would like this test to fail if for some reason we revert back to byte-wise implementations, right?

File: sw/device/lib/base/memory_perftest.c
Comment (alphan) [reply sw/device/lib/base/memory_perftest.c:36]: Please ignore my comment :) For some reason I thought we would have separate setup functions for each buffer (`setup_bufN()` would setup `context->bufN` using `context->len`) but it doesn't make sense :).

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:46]: Well, it depends on the nature of the perftest, which I haven't quite nailed down yet. Currently memcpy, memset, memcmp, and memrcmp have zero variance. The others have too much to fit in a u32, so I can't easily print the results with `base_printf()`! For instance, the memchr perftest randomizes `buf1` and `buf2` and then searches `buf1` for `buf2[0]`.

I'm happy to just delete the stats computation if we land on a suite of perftests that always have zero variance. For now, I'm leaning on that to have a sense of how deterministic the tests are :)

**Comment by alphan** [comment]
@cfrantz, are you ok with introducing the new `_perftest` suffix or should we call this one `_functest` as well?

**Comment by dmcardle** [comment]
Partially responded to feedback. Thanks, @cfrantz and @alphan!

FYI, this is what the output looks like:

```
I00000 test_rom.c:88] Version: earlgrey_silver_release_v5-6540-g09115123e, Build Date: 2022-07-28 12:43:17
I00001 test_rom.c:105] TestROM:e3ace18d
I00000 test_rom.c:88] Version: earlgrey_silver_release_v5-6540-g09115123e, Build Date: 2022-07-28 12:43:17
I00001 test_rom.c:105] TestROM:e3ace18d
I00002 test_rom.c:137] Test ROM complete, jumping to flash!
I00000 ottf_main.c:80] Running sw/device/lib/base/memory_perftest.c
memcpy                     10515565 cycles (mean)                    0 cycles (variance)                    4 (n)
memcpy_zeroes              10515565 cycles (mean)                    0 cycles (variance)                    4 (n)
memset                      6523553 cycles (mean)                    0 cycles (variance)                    4 (n)
memcmp                     10513567 cycles (mean)                    0 cycles (variance)                    4 (n)
memrcmp                    10017063 cycles (mean)                    0 cycles (variance)                    4 (n)
W00001 memory_perftest.c:293] Cannot print statistics for memchr because mean and/or variance do not fit in 32 bits
W00002 memory_perftest.c:293] Cannot print statistics for memrchr because mean and/or variance do not fit in 32 bits
I00003 ottf_main.c:56] Finished sw/device/lib/base/memory_perftest.c
I00004 status.c:28] PASS!
```

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:302]: Oh! I thought you were talking about checking input/output expectations. Are you talking about checking runtime expectations? Or both?

Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:302]: Bizarre... I tried encoding the expected number of cycles and setting `num_reps=1`. I set the expected cycle count to 9020065, but when I recompile and run, the memcpy test took 10019061 cycles. OK, I updated the expectation, but now it took 10515565 cycles. But that's apparently a fixed point, now it consistently takes 10515565 cycles.

Between recompiling, I can rerun the memcpy test any number of times and it's always has the same cycle count as the last run. It seems like editing the source and rerunning the test somehow influences the runtime. Maybe pushing the bitstream to the FPGA re-seeds the RNG or something?

Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:302]: I replaced the definition of `fill_buf_arbitrary_values()` with something deterministic and now the cycle count is also deterministic across recompiles. Huh.

**Comment by dmcardle** [comment]
Just pushed another commit that simplifies things. PTAL! It removes the variance computation and hardcodes expected cycle counts. When the actual cycle count doesn't match the expectation, it fails the test. I'm slightly worried this will be brittle and break when we update the compiler, etc., but I'm also happy to see what happens!

**Comment by dmcardle** [comment]
> Just pushed another commit that simplifies things. PTAL! It removes the variance computation and hardcodes expected cycle counts. When the actual cycle count doesn't match the expectation, it fails the test. I'm slightly worried this will be brittle and break when we update the compiler, etc., but I'm also happy to see what happens!

@milesdai recommended a ratcheting mechanism where you only fail the test when the performance gets worse. I think that's at least worth keeping in our back pocket, if  this test turns out to be flaky.

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:57]: Although we should never see the counters wrap, the following subtract (`end - start`) is well defined and would result in  the correct cycle count for `uint64_t` args (or for any (start,end) of the same unsigned type).

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:196]: Any particular reason to put these on the stack and initialize them? (ie: if they were uninitialized globals, they wouldn't have to be constrained by some notion of stack size and they'd be guaranteed to be zero-initialized because they'd be in the `.bss` section).

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [inline sw/device/lib/base/memory_perftest.c:137]: (5) The icache gets turned on prior to test execution.

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [reply sw/device/lib/base/memory_perftest.c]: The SRAM is a 32-bit memory plus ECC bits.  Writing to it in 8-bit or 16-bit chunks causes a read-modify-write behind the scenes, along with recalculating the ECC bits.  There is also scrambling of the RAM so that an attacker with access to the die can't just read out RAM to get the values.

Which RNG were you using to generate random values?  `rand_testutils_gen32` should be deterministic in its outputs, _however_, from run-to-run, its timing may vary because of the internal construction of that routine.

**Comment by cfrantz** [comment]
> Just pushed another commit that simplifies things. PTAL! It removes the variance computation and hardcodes expected cycle counts. When the actual cycle count doesn't match the expectation, it fails the test. I'm slightly worried this will be brittle and break when we update the compiler, etc., but I'm also happy to see what happens!

I'm also concerned that encoding exact cycle counts may be brittle, but I don't have a better suggestion at the moment.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c]: `rand_testutils_gen32()` is what I was using. I didn't squash commits yet, so you can still see that version here: https://github.com/lowRISC/opentitan/pull/13871/commits/9b74b9b07eb029030f84daa8601122f39b67a2ef

Aha, so it sounds like scrambling was taking an unpredictable number of cycles? That would certainly explain why hardcoding the cycle-count expectations was not working.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:137]: Thanks!

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [reply sw/device/lib/base/memory_perftest.c:57]: I'd prefer to, but I also can't ever see a situation where the CHECK would fail.

File: sw/device/lib/base/memory_perftest.c
Comment (cfrantz) [reply sw/device/lib/base/memory_perftest.c]: I don't think it was the scrambling or ECC calcuation (that was my first thought, but after thinking about it for a bit, I decided that it can't possibly be the hardware causing the cycle variance).

It's `rand_testutils_gen32` that was causing the variance.  That routine isn't reseeded between runs, and the main bit of that routine that creates the random output is:
```
if (lsb) lfsr ^= kConst
```
Since it isn't reseeded between runs, the stream of outputs will be different from run to run and thus the number of times the  internal "mixing" branch is taken will be different from run to run.

This probably could be solved by changing that `if` statement into something like:
```
lfsr ^= lsb ? kConst : 0;
```
But even then, we'd have to look at how the compiler generated code for that conditional.  We'd probably have to use the constant time routines from `hardened.h`:
```
lfsr ^= ct_cmov32(ct_seqz32(lsb), 0, kConst)
```

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c:57]: Sounds good, I'll do that in the next patchset. The CHECK is not protecting us from UB, nor is it preserving correctness if we believe that the runtime of a single test will never exceed 2**64 cycles, which seems pretty believable given my THz estimate.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c]: Resolving this because I'm not using `rand_testutils_gen32()` anymore.

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c]: Resolving this — no need to keep track of the code for computing variance, it will live on in my memory :P

File: sw/device/lib/base/memory_perftest.c
Comment (dmcardle) [reply sw/device/lib/base/memory_perftest.c]: Resolving — I have since deleted the stats computation. Let's go ahead with these hardcoded expectations on cycle counts until we see whether it's too flaky.

**Comment by dmcardle** [comment]
Removed the CHECK and updated the comment explaining why `end_cycles - start_cycles` is well-defined and correct. PTAL!

**Review by alphan — APPROVED** [review]
Thanks @dmcardle! This test will probably fail some time soon but I think it would be interesting to see the change. After that we can relax the expectation to a range.

**Comment by dmcardle** [comment]
Awesome! Thanks for the detailed review, folks! I'm marking this ready to merge, unless @cfrantz has more thoughts.

**Comment by alphan** [comment]
@dmcardle Can you squash commits and push again? I think a single commit should be enough.

**Comment by alphan** [comment]
Created a [gist](https://gist.github.com/alphan/68830dc4672aed770855e68228d5dade) for the variance computation in case we want to bring it back. Thanks @dmcardle!

**Comment by dmcardle** [comment]
> @dmcardle Can you squash commits and push again? I think a single commit should be enough.

Done! Also update the commit message — it still mentioned printing statistics to stdout.

**Comment by dmcardle** [comment]
Two things:

* Looks like the new test is failing on CI, but I'm not sure how to grab that test-specific log file. https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=88879&view=logs&j=be0b5a13-ded6-5a2a-3e48-80a10e7eef12&t=2aaa9e8b-9b00-5802-e0c1-7bc55418740f&l=24612

* I realized that when I've been running the perf test locally, I've been specifying `-O2`. CI is probably not using the same optimization level. Is there a reason that we don't specify `-O2` in `.bazelrc`?

**Comment by dmcardle** [comment]
Latest patch has hardcoded expectations *not* based on `-O2`. I just ran `./bazelisk.sh test --cache_test_results=no --test_output=all //sw/device/lib/base:memory_perftest_fpga_cw310` and updated the expectations based on stdout.

**Pull request merged**
